mongo/integration/client_test.go

```diff
@@ -597,6 +598,63 @@ func TestClient(t *testing.T) {
 		closed := len(tpm.Events(func(e *event.PoolEvent) bool { return e.Type == event.ConnectionClosed }))
 		assert.Equal(t, 0, closed, "expected no connections to be closed")
 	})
+
+	// Test that OP_MSG is used for authentication-related commands on 3.6+ (WV 6+). Do not test when API version is
+	// set, as handshakes will always use OP_MSG.
+	opMsgOpts := mtest.NewOptions().ClientType(mtest.Proxy).MinServerVersion("3.6").Auth(true).RequireAPIVersion(false)
+	mt.RunOpts("OP_MSG used for authentication on 3.6+", opMsgOpts, func(mt *mtest.T) {
+		err := mt.Client.Ping(context.Background(), mtest.PrimaryRp)
+		assert.Nil(mt, err, "Ping error: %v", err)
+
+		msgPairs := mt.GetProxiedMessages()
+		assert.True(mt, len(msgPairs) >= 2, "expected at least 2 events, got %v", len(msgPairs))
+
+		// First two messages should be connection handshakes: one for the heartbeat connection and the other for the
+		// application connection. These handshakes should use OP_QUERY as their OpCode, as wire version is not yet
+		// known.
+		for idx, pair := range msgPairs[:2] {
+			assert.Equal(mt, internal.LegacyHello, pair.CommandName, "expected command name %s at index %d, got %s",
+				internal.LegacyHello, idx, pair.CommandName)
+
+			// Assert that appended OpCode is OP_QUERY.
+			assert.Equal(mt, wiremessage.OpQuery, pair.Sent.OpCode,
+				"expected 'OP_QUERY' OpCode in wire message, got %q", pair.Sent.OpCode.String())
+		}
+
+		// Look for a saslContinue in the remaining proxied messages and assert that it uses the OP_MSG OpCode, as wire
+		// version is now known to be >= 6.
+		var saslContinueFound bool
+		for _, pair := range msgPairs[2:] {
+			if pair.CommandName == "saslContinue" {
+				saslContinueFound = true
+				assert.Equal(mt, wiremessage.OpMsg, pair.Sent.OpCode,
+					"expected 'OP_MSG' OpCode in wire message, got %s", pair.Sent.OpCode.String())
+				break
+			}
+		}
+		assert.True(mt, saslContinueFound, "did not find 'saslContinue' command in proxied messages")
+	})
+
+	// Test that OP_MSG is used for handshakes when API version is declared.
+	opMsgSAPIOpts := mtest.NewOptions().ClientType(mtest.Proxy).MinServerVersion("3.6").RequireAPIVersion(true)
```

Comment:
```suggestion
	opMsgSAPIOpts := mtest.NewOptions().ClientType(mtest.Proxy).MinServerVersion("5.0").RequireAPIVersion(true)
```

Versioned API only applies to MongoDB 5.0+ servers.